From 54b4327565b5371d7ffef7c17b8de587c037909d Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Tue, 15 Dec 2015 04:45:56 -0800
Subject: [PATCH 0874/1767] UBUNTU: SAUCE: fixup: 20/23 locking issue around in
 __label_update

aa_label_remove needs to take the ls write lock but, we are already
holding it so move to __aa_label_remove

also fix bug where the old label is not being forwarded to the new label
when it results in a new namespace due to an the old ns being removed.

Signed-off-by: John Johansen <john.johansen@canonical.com>
Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 security/apparmor/label.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/security/apparmor/label.c b/security/apparmor/label.c
index ee9a573..0b3a80a 100644
--- a/security/apparmor/label.c
+++ b/security/apparmor/label.c
@@ -1844,7 +1844,7 @@ static struct aa_label *__label_update(struct aa_label *label)
 	if (invcount) {
 		l->size -= aa_sort_and_merge_profiles(l->size, &l->ent[0]);
 		if (labels_set(label) != labels_set(l)) {
-			aa_label_remove(labels_set(label), label);
+			__aa_label_remove(labels_set(label), label, l);
 			write_unlock_irqrestore(&ls->lock, flags);
 			tmp = aa_label_insert(labels_set(l), l);
 			goto out;
-- 
1.9.3

